### PR TITLE
[addons][vfs] cleanup API, improve "C" interface, not use void* direct as value

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

This thought to make in way equal to PVR API, so to have the "C" in separate header and helper class.

Commit 1:
-----------------------------------
**[addons][vfs][filesystem] increase stat support and use also on VFS addon**

There becomes the following stat parts support added to STAT_STRUCTURE:
- isBlock (S_IFBLK)
- isCharacter (S_IFCHR)
- isFifo (S_IFIFO)
- isRegular (S_IFREG)
- isSocket (S_IFSOCK)

And then also with STAT_STRUCTURE/FileStatus used on VFS addon stat calls.

Commit 2:
-----------------------------------
**[addons][vfs] use C++ kodi::addon::VFSUrl for URL**

Before was the "C" structure used, this change to use the C++ translation class about, done to match the with PVR started style about.

Commit 3:
-----------------------------------
**[addons][vfs] improve IO file calls**

There was before on few IO types a void* used to write/read where makes it harder for other languages and more unsafe in C/C++.

This change every for addons usable part in a own function where the void* no more needed.

Commit 4:
-----------------------------------
**[addons][vfs] prevent direct use of `"void*"` on addon interface**

There becomes the "void*" usage replaced to a "C define VFS_FILE_HANDLE" and "C++ using VFSFileHandle".
Also is a from addon used memory pointer changed from void* to uint8_t*. 
On other parts where only relate to Kodi is changed to KODI_HANDLE.

So is on addon not direct a void* used and more clean on other languages.

Commit 5:
-----------------------------------
**[addons][vfs] separate "C" and "C++" to different headers**

This place the "C" parts to "kodi/c-api/addon-instance/vfs.h" to have
clean for other languages and to confirm a correct ABI between Kodi
and addon used.

Commit 6:
-----------------------------------
**[addons][vfs] cleanup vfs addon instance documentation**

To make it a bit cleaner, fix few doxygen parts and add few new text parts.

Commit 7:
-----------------------------------
**[addons][filesystem] allow recursive directory delete on kodi::vfs::RemoveDirectory**

There becomes another "C" part added about (to prevent min version change")
where then called if kodi::vfs::RemoveDirectory(..., true).

Commit 8:
-----------------------------------
**[addons][filesystem][vfs] increase API versions**

There becomes increased:
- ADDON_INSTANCE_VERSION_VFS = 3.0.0
- ADDON_INSTANCE_VERSION_VFS_MIN = 3.0.0
- ADDON_GLOBAL_VERSION_FILESYSTEM = 1.1.5

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/VFSEntry.cpp [hunk 1]
```diff
@@ -281,47 +330,62 @@ int64_t CVFSEntry::GetLength(void* ctx)
...
-      kodiData->maxrate = data.maxrate;
-      kodiData->currate = data.currate;
-      kodiData->lowspeed = data.lowspeed;
+      if (!m_struct.toAddon->io_control_set_cache_rate)
+        return -1;
+
+      return m_struct.toAddon->io_control_set_cache_rate(&m_struct, ctx,
+                                                         *static_cast<unsigned int*>(param))
+                 ? 1
+                 : 0;
```
Comment (garbear) [inline xbmc/addons/VFSEntry.cpp:369]: The clang formatting of this line is a little awkward. If you alias `param` you could probably fit everything on one line:

```c++
unsigned int& iParam = *static_cast<unsigned int*>(param);
```

Comment (AlwinEsch) [reply xbmc/addons/VFSEntry.cpp]: Thanks, with them looks much cleaner, have changed to them.

**Comment by garbear** [comment]
git_commit

**Comment by AlwinEsch** [comment]
The "burn my head away" commits are removed now :grinning:.

Yeah, your image match exact how my head works :+1: 

From my side should it now finished and on all related addons requests about and runtime tests OK.

**Pull request merged**